<html><head><meta charset='utf-8'><meta name='viewport' content='width=device-width, initial-scale=1'>
<meta name='applicable-device' content='pc'><meta name='keywords' content='电脑,电脑讲解,电脑技术,编程,电脑故障维修C++中的异常处理（上）' />
<script src='../../highlight/highlight.pack.js'></script>
<link rel='stylesheet' type='text/css' href='../../highlight/styles/monokai.css'/>

<link rel='stylesheet' href='../../fenxiang/dist/css/share.min.css'>
<script src='../../fenxiang/src/js/social-share.js'></script>
<script src='../../fenxiang/src/js/qrcode.js'></script>

</head><body><script>hljs.initHighlightingOnLoad();</script><script>
var system ={};  
var p = navigator.platform;       
system.win = p.indexOf('Win') == 0;  
system.mac = p.indexOf('Mac') == 0;  
system.x11 = (p == 'X11') || (p.indexOf('Linux') == 0);     
if(system.win||system.mac||system.xll){
document.write("<link href='../css/3.css' rel='stylesheet' type='text/css'>");}else{ document.write("<link href='../css/3wap.css' rel='stylesheet' type='text/css'>");}</script><script src='../../js/3.js'></script><div class='div2'><div class='heading_nav'><ul><div><li><a href='../../index.html'>首页</a></li>
</div><div onclick='hidden1()' >分享</div>
</ul></div></div>
<div id='heading_nav2'> 
<li class='row' >
<div class='social-share' data-mode='prepend'><a href='javascript:' class='social-share-icon icon-heart'></a></div></li></div><script charset='utf-8' src='../../3/js/hengfu.js'></script><script charset='utf-8' src='../../3/js/hengfu2.js'></script><hr><div class='div1'><div class='biaoti'><center>C++中的异常处理（上）</center></div><div class='banquan'>原文出处:本文由博客园博主子宇24提供。<br/>
原文连接:https://www.cnblogs.com/dishengAndziyu/p/10919819.html</div><br>
    <p align="left">1，C++ 内置了异常处理的语法元素 try ... catch ...；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，try 语句处理正常代码逻辑；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，catch 语句处理异常情况；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，try 语句中的异常由对应的 catch 语句处理；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，代码示例：</p>
<div class="cnblogs_code">
<pre><code><span style="color: #008080;">1</span> <span style="color: #0000ff;">try</span>
<span style="color: #008080;">2</span> <span style="color: #000000;">{
</span><span style="color: #008080;">3</span>     <span style="color: #0000ff;">double</span> r = divide(<span style="color: #800080;">1</span>, <span style="color: #800080;">0</span>);  <span style="color: #008000;">//</span><span style="color: #008000;"> divide() 函数里面产生除 0 异常，于是 divide() 函数就会将这个异常抛出到调用的地点，这个地点正好在 try 语句块中，其会将捕捉到的异常扔给下面的 catch 代码块，catch 代码块就可以将除 0 异常捕捉到，于是打印 Divided by zero ...；</span>
<span style="color: #008080;">4</span> <span style="color: #000000;">}
</span><span style="color: #008080;">5</span> <span style="color: #0000ff;">catch</span><span style="color: #000000;">(...)
</span><span style="color: #008080;">6</span> <span style="color: #000000;">{
</span><span style="color: #008080;">7</span>     cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">Divided by zero ...</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">8</span> }</pre>
</div>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;</p>
<p align="left">2，C++ 通过 throw 语句抛出异常信息：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，代码示例：</p>
<div class="cnblogs_code">
<pre><code><span style="color: #008080;"> 1</span> <span style="color: #0000ff;">double</span> divide(<span style="color: #0000ff;">double</span> a, <span style="color: #0000ff;">double</span><span style="color: #000000;"> b)
</span><span style="color: #008080;"> 2</span> <span style="color: #000000;">{
</span><span style="color: #008080;"> 3</span>     <span style="color: #0000ff;">const</span> <span style="color: #0000ff;">double</span> delta = <span style="color: #800080;">0.000000001</span><span style="color: #000000;">;
</span><span style="color: #008080;"> 4</span>     <span style="color: #0000ff;">double</span> ret = <span style="color: #800080;">0</span><span style="color: #000000;">;
</span><span style="color: #008080;"> 5</span>            
<span style="color: #008080;"> 6</span>     <span style="color: #0000ff;">if</span>( !((-delta &lt; b) &amp;&amp; (b &lt;<span style="color: #000000;"> delta)) )
</span><span style="color: #008080;"> 7</span> <span style="color: #000000;">    {
</span><span style="color: #008080;"> 8</span>            ret = a /<span style="color: #000000;"> b;
</span><span style="color: #008080;"> 9</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">10</span>     <span style="color: #0000ff;">else</span>
<span style="color: #008080;">11</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">12</span>            <span style="color: #0000ff;">throw</span> <span style="color: #800080;">0</span>;  <span style="color: #008000;">//</span><span style="color: #008000;"> 0 这个字面值常量代表当前的异常元素，表示产生除 0  异常； 异常元素可以是一个值、一个对象、一个字符串等； 扔出异常元素后，divide() 就会立即返回到调用点；</span>
<span style="color: #008080;">13</span> <span style="color: #000000;">     }
</span><span style="color: #008080;">14</span>            
<span style="color: #008080;">15</span>      <span style="color: #0000ff;">return</span><span style="color: #000000;"> ret;
</span><span style="color: #008080;">16</span> }</pre>
</div>
<p align="left">&nbsp;&nbsp;&nbsp; 2，C++ 中 throw 这个新的关键字就是用来仍出异常元素的，我们简称为&ldquo;异常&rdquo;，每一个异常都需要被 catch 语句来处理；</p>
<p align="left">&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;</p>
<p align="left">3，C++ 异常处理分析：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，throw 抛出的异常必须被 catch 处理：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，当前函数能够处理异常，程序往下执行；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，当前函数无法处理异常，则函数停止执行，并返回；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，返回是异常返回，和正常返回不一样；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，异常产生后的函数返回是没有返回值的；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，未被处理的异常会顺着函数调用栈向上传播，知道被处理为止，否则程序将停止执行；</p>
<p align="center"><img src="./images/C++中的异常处理（上）0.png" alt="" /></p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，C++ 异常处理初探编程实验：</p>
<div class="cnblogs_code">
<pre><code><span style="color: #008080;"> 1</span> #include &lt;iostream&gt;
<span style="color: #008080;"> 2</span> #include &lt;<span style="color: #0000ff;">string</span>&gt;
<span style="color: #008080;"> 3</span> 
<span style="color: #008080;"> 4</span> <span style="color: #0000ff;">using</span> <span style="color: #0000ff;">namespace</span><span style="color: #000000;"> std;
</span><span style="color: #008080;"> 5</span> 
<span style="color: #008080;"> 6</span> <span style="color: #0000ff;">double</span> divide(<span style="color: #0000ff;">double</span> a, <span style="color: #0000ff;">double</span><span style="color: #000000;"> b)
</span><span style="color: #008080;"> 7</span> <span style="color: #000000;">{
</span><span style="color: #008080;"> 8</span>     <span style="color: #0000ff;">const</span> <span style="color: #0000ff;">double</span> delta = <span style="color: #800080;">0.000000000000001</span><span style="color: #000000;">;
</span><span style="color: #008080;"> 9</span>     <span style="color: #0000ff;">double</span> ret = <span style="color: #800080;">0</span><span style="color: #000000;">;
</span><span style="color: #008080;">10</span>     
<span style="color: #008080;">11</span>     <span style="color: #0000ff;">if</span>( !((-delta &lt; b) &amp;&amp; (b &lt;<span style="color: #000000;"> delta)) )
</span><span style="color: #008080;">12</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">13</span>         ret = a /<span style="color: #000000;"> b;
</span><span style="color: #008080;">14</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">15</span>     <span style="color: #0000ff;">else</span>
<span style="color: #008080;">16</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">17</span>         <span style="color: #0000ff;">throw</span> <span style="color: #800080;">0</span><span style="color: #000000;">;  
</span><span style="color: #008080;">18</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">19</span>     
<span style="color: #008080;">20</span>     <span style="color: #0000ff;">return</span><span style="color: #000000;"> ret;
</span><span style="color: #008080;">21</span> <span style="color: #000000;">}
</span><span style="color: #008080;">22</span> 
<span style="color: #008080;">23</span> <span style="color: #0000ff;">int</span> main(<span style="color: #0000ff;">int</span> argc, <span style="color: #0000ff;">char</span> *<span style="color: #000000;">argv[])
</span><span style="color: #008080;">24</span> <span style="color: #000000;">{    
</span><span style="color: #008080;">25</span>     <span style="color: #0000ff;">try</span>
<span style="color: #008080;">26</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">27</span>         <span style="color: #0000ff;">double</span> r = divide(<span style="color: #800080;">1</span>, <span style="color: #800080;">0</span>);  <span style="color: #008000;">//</span><span style="color: #008000;"> 扔出异常后，divide() 自身没有 try...catch...，不能处理异常，所以 divide() 这个函数就会立即的停止并且带着异常元素返回到调用点，而 main() 函数中当也没有try...catch... 语句时，异常向上传播，main() 已经是第一个函数了，则导致整个程序异常停止，编译器显示：terminate called after throwing an instance of 'int'，已放弃；当 main() 中包含 try...catch... 语句时，程序正常打印：Divided by zero...；</span>
<span style="color: #008080;">28</span>             
<span style="color: #008080;">29</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">r = </span><span style="color: #800000;">"</span> &lt;&lt; r &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">30</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">31</span>     <span style="color: #0000ff;">catch</span>(...)  <span style="color: #008000;">//</span><span style="color: #008000;"> &ldquo;...&rdquo; 表示见下分析；</span>
<span style="color: #008080;">32</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">33</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">Divided by zero...</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">34</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">35</span>     
<span style="color: #008080;">36</span>     <span style="color: #0000ff;">return</span> <span style="color: #800080;">0</span><span style="color: #000000;">;
</span><span style="color: #008080;">37</span> }</pre>
</div>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p>
<p align="left">4，同一个 try 语句可以跟上多个 catch 语句：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，catch 语句可以定义具体处理的异常类型；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，不同类型的异常由不同的 catch 语句负责处理；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，try 语句中可以抛出任何类型的异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，try 语句块用于放置可能产生异常的代码，工程里面一般都习惯将正常功能逻辑代码放在 try 语句块中，然后后面跟上多个 catch 语句；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，可以是一个 int、字符串、对象；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 4，catch(...) 用于处理所有类型的异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，一般情况作为最后出场的语句块；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 5，任何异常都只能被捕获（catch）一次；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，只要被 catch 一次，其它的 catch 就没有捕获机会了；</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">5，异常处理的匹配规则：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，示意图：</p>
<p align="center">&nbsp;<img src="./images/C++中的异常处理（上）1.png" alt="" /></p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，try 语句块中直接通过 throw 扔出整型异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，扔出后发现是在 catch 语句块中，然后这个异常就会进行 catch 语句块的匹配；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 4，从上向下严格匹配每个 catch 它所需要捕获的异常类型；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 5，匹配 catch 语句块所标识的类型时，是不会进行任何的类型转换的；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 6，如果说不能够严格匹配上，那么当前的异常没有被捕获，当前的函数就会立即的停止执行，并且带着这个异常沿着函数调用栈顺序的返回；</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">6，异常类型匹配编程实验：</p>
<div class="cnblogs_code">
<pre><code><span style="color: #008080;"> 1</span> #include &lt;iostream&gt;
<span style="color: #008080;"> 2</span> #include &lt;<span style="color: #0000ff;">string</span>&gt;
<span style="color: #008080;"> 3</span> 
<span style="color: #008080;"> 4</span> <span style="color: #0000ff;">using</span> <span style="color: #0000ff;">namespace</span><span style="color: #000000;"> std;
</span><span style="color: #008080;"> 5</span> 
<span style="color: #008080;"> 6</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> Demo1()
</span><span style="color: #008080;"> 7</span> <span style="color: #000000;">{
</span><span style="color: #008080;"> 8</span>     <span style="color: #0000ff;">try</span>
<span style="color: #008080;"> 9</span> <span style="color: #000000;">    {   
</span><span style="color: #008080;">10</span>         <span style="color: #0000ff;">throw</span> <span style="color: #800000;">'</span><span style="color: #800000;">c</span><span style="color: #800000;">'</span><span style="color: #000000;">;
</span><span style="color: #008080;">11</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">12</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">char</span><span style="color: #000000;"> c)
</span><span style="color: #008080;">13</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">14</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(char c)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">15</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">16</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">short</span><span style="color: #000000;"> c)
</span><span style="color: #008080;">17</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">18</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(short c)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">19</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">20</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">double</span><span style="color: #000000;"> c)
</span><span style="color: #008080;">21</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">22</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(double c)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">23</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">24</span>     <span style="color: #0000ff;">catch</span>(...)  <span style="color: #008000;">//</span><span style="color: #008000;"> 当把它放到其它 catch 语句块前面的时候，编译器显示：error: '...' handler must be the last handler for its try block；</span>
<span style="color: #008080;">25</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">26</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(...)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">27</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">28</span> <span style="color: #000000;">}
</span><span style="color: #008080;">29</span> 
<span style="color: #008080;">30</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> Demo2()
</span><span style="color: #008080;">31</span> <span style="color: #000000;">{
</span><span style="color: #008080;">32</span>     <span style="color: #0000ff;">throw</span> <span style="color: #800000;">"</span><span style="color: #800000;">D.T.Software;  // 两个分开执行的；</span>
<span style="color: #008080;">33</span>     <span style="color: #0000ff;">throw</span> <span style="color: #0000ff;">string</span>(<span style="color: #800000;">"</span><span style="color: #800000;">D.T.Software</span><span style="color: #800000;">"</span><span style="color: #000000;">);
</span><span style="color: #008080;">34</span> <span style="color: #000000;">}
</span><span style="color: #008080;">35</span> 
<span style="color: #008080;">36</span> <span style="color: #0000ff;">int</span> main(<span style="color: #0000ff;">int</span> argc, <span style="color: #0000ff;">char</span> *<span style="color: #000000;">argv[])
</span><span style="color: #008080;">37</span> <span style="color: #000000;">{    
</span><span style="color: #008080;">38</span> <span style="color: #000000;">    Demo1();
</span><span style="color: #008080;">39</span>     
<span style="color: #008080;">40</span>     <span style="color: #0000ff;">try</span>
<span style="color: #008080;">41</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">42</span>         Demo2();  <span style="color: #008000;">//</span><span style="color: #008000;"> 认为它可能产生字符串异常，所以将它放到了这里；                 </span>
<span style="color: #008080;">43</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">44</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">char</span>*<span style="color: #000000;"> s)
</span><span style="color: #008080;">45</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">46</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(char *s)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">47</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">48</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">const</span> <span style="color: #0000ff;">char</span>* cs)  <span style="color: #008000;">//</span><span style="color: #008000;"> 没有这个 catch 语句块时，第一个异常执行后编译器显示：terminate called after throwing an instance of 'char const *'；</span>
<span style="color: #008080;">49</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">50</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(const char *cs)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">51</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">52</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">string</span><span style="color: #000000;"> ss)
</span><span style="color: #008080;">53</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">54</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(string ss)</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">55</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">56</span>     
<span style="color: #008080;">57</span>     <span style="color: #0000ff;">return</span> <span style="color: #800080;">0</span><span style="color: #000000;">;
</span><span style="color: #008080;">58</span> }</pre>
</div>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">7，小结：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，C++ 中直接支持异常处理的概念；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，try ... catch ... 是 C++ 中异常处理的专用语句；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，try 语句处理正常代码逻辑，catch 语句处理异常情况；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 4，同一个 try 语句可以跟上多个 catch 语句；</p>
<p>&nbsp;&nbsp;&nbsp; 5，异常处理必须严格匹配，不进行任何的类型转换；</p>
</div>
</div><hr><script charset='utf-8' src='../../js/sming.js'></script></body></html>